### ISSUE TITLE
wrong coordinate when connecting DIVs generated by javascript dynamically 

### ISSUE THREAD

**Issue by xu4wang** [jsplumb/jsplumb#1131]
Hello 

I have some code generating DIV as below.  I set the left and top style so the DIV will be located in the proper place on the canvas.

```javascript
class Node {
  constructor(id, title, parent) {
    this.id = id;
    var childNode = document.createElement('div');
    childNode.innerHTML = title;
    childNode.className = 'window';
    childNode.id = id;
    this.ele = childNode;
    var canvas = document.getElementById(parent);
    canvas.appendChild(childNode);
  }

  layout(row, col, width, height) {
    //calculate the upper,left position
    var top = height * (row - 1 );
    var left = width * (col - 1 );
    this.ele.style.top = top + 'px';
    this.ele.style.left = left + 'px';
    this.ele.style.width = width + 'px';
    this.ele.style.height = height + 'px';
  }
}
```

And then, I use connect method to connect the DIVs created.

```javascript
function add_edge(attrs) {
  attrs.source = id2element(attrs.source);
  attrs.target = id2element(attrs.target);
  var instance = window.j;
  instance.connect(attrs);
}
```

Checked the log and the actual instance.connect is with below parameters.  same as examples in https://docs.jsplumbtoolkit.com/community/lib/overlays

```javascript 
instance.connect({
    source:someElement,
    target:someOtherElement,
    anchor: "AutoDefault",
    endpoints: ["Dot", "Blank"], 
    overlays:[ 
        { type:"Arrow", options:{location:1}},
        { 
            type:"Label", 
            options:{ label:"foo", location:50, id:"myLabel" } 
        }
    ]
})
```

However, the generated link, the svg object, is with wrong position and wrong size.

```xml
<svg style="" width="0" height="0" pointer-events="none" position="absolute" version="1.1" xmlns="http://www.w3.org/2000/svg" class="jtk-connector"><path d="M 0 0 L 180 0 " transform="translate(0,0)" pointer-events="visibleStroke" version="1.1" xmlns="http://www.w3.org/2000/svg" fill="none" stroke="#666"></path><path jtk-overlay-id="984138e6-c4e8-4522-9549-cd0e3241acb5" version="1.1" xmlns="http://www.w3.org/2000/svg" class="jtk-overlay" d="M180,0 L160,10 L167.54,0 L160,-10 L180,0" stroke="#666" fill="#666" transform="translate(0,0)" pointer-events="visibleStroke"></path></svg>

```
It look like the link is considering the two DIVs are all located in (0,0).  However, I've dynamically change the left/top/width/height style via javascript before instance.connect is called.

PS. The label "foo" in below screen capture are displayed correctly as expected. 
<img width="672" alt="image" src="https://user-images.githubusercontent.com/311397/171373926-b005d01d-0274-46a8-9197-6e902e7eebfc.png">

Is there something I'm missing?

Thanks,Austin

**Comment by xu4wang** [comment]
In summary:

1. the connection itself and the arrow overlay are with wrong position and size.
2. the label overlay is with correct position.
3. The endpoint is shown at the correct position.

**Comment by xu4wang** [comment]
digging into the other issues,  it looks like I need to call

```javascript
  instance.manage(ele);
```

after the new DIV is created dynamically.  I added that.  The DIV become draggable. However the arrow connection is still not showing.

**Comment by sporritt** [comment]
you dont need to call manage yourself. connect will call manage.

are you sure the node's layout method gets called before the connect call?

**Comment by xu4wang** [comment]
https://github.com/xu4wang/diagram/tree/main/src/template

**Comment by xu4wang** [comment]
> you dont need to call manage yourself. connect will call manage.
> 
> are you sure the node's layout method gets called before the connect call?

Yes. layout was called first,  before connect.

**Comment by xu4wang** [comment]
I have successfully re-produce this issue using the /demo/draggableConnectors coming with the standard latest jsplumb.

I created two chrome snippets.

*  addnodes will add two DIVs into canvas.

```addnodes
        var childNode = document.createElement('div');
        childNode.innerHTML = 'Node1';
        childNode.className = 'window';
        childNode.id = 'n1';
        var canvas = document.getElementById('canvas');
        canvas.appendChild(childNode);
        childNode.style.top = 100 + 'px';
        childNode.style.left = 100 + 'px';

        var childNode2 = document.createElement('div');
        childNode2.innerHTML = 'Node2';
        childNode2.className = 'window';
        childNode2.id = 'n2';
        canvas.appendChild(childNode2);
        childNode2.style.top = 300 + 'px';
        childNode2.style.left =300 + 'px';
```
* connect will connect the two nodes added.

```javascript
window.j.connect({
    source:childNode,
    target:childNode2,
    anchor: "AutoDefault",
    endpoints: ["Dot", "Blank"], 
    overlays:[ 
        { type:"Arrow", options:{location:1}},
        { 
            type:"Label", 
            options:{ label:"foo", location:50, id:"myLabel" } 
        }
    ]
})
```

After running the standard draggable connectors demo, the two snippets will be executed one by one manually.

Please find below the screen capture. 

<img width="1417" alt="image" src="https://user-images.githubusercontent.com/311397/171781921-04d24aa4-8283-41da-9000-3c21c28e2faf.png">

1. the connection itself and the arrow overlay are with wrong position and size.
2. the label overlay is with correct position.
3. The endpoint is shown at the correct position.

**Comment by sporritt** [comment]
this isn't a jsplumb bug, it's an issue in the package.json for that demo

https://github.com/jsplumb-demonstrations/draggable-connections/issues/2

for anyone who comes across this issue in the future, it is imperative that you have the same version of every package installed. You cannot use, say, 5.6.5 bezier connector with 5.6.0 browser-ui.

**Comment by xu4wang** [comment]
I am not sure if this is the issue. I just update the libraries to use:

```
  "dependencies": {
    "@jsplumb/browser-ui": "^5.0.0",
    "@jsplumb/connector-bezier": "^5.0.0"
  }
```

still the same result.

**Comment by xu4wang** [comment]
> this isn't a jsplumb bug, it's an issue in the package.json for that demo
> 
> [jsplumb-demonstrations/draggable-connections#2](https://github.com/jsplumb-demonstrations/draggable-connections/issues/2)
> 
> for anyone who comes across this issue in the future, it is imperative that you have the same version of every package installed. You cannot use, say, 5.6.5 bezier connector with 5.6.0 browser-ui.

One more question, actually I was using the demo inside jsplumb repo, say 

```
/github/jsplumb/demo/draggableConnectors
```

it is using the libraries from 
```
/github/jsplumb/dist
```

I believe ``` /github/jsplumb/dist ``` are all for the current version? I checked the package.json for each package in dist, they are all 5.7.1

**Comment by xu4wang** [comment]
Updated jsplumb to 5.7.2, still no luck. 

Please help re-open this issue.   @sporritt

**Comment by sporritt** [comment]
there's a lot of to and fro on this issue now, so it's time for some bullet points. you can never go wrong with bullet points.

- the draggable connector demo on our site is deployed from the other repository. it is not the demo from inside this repository. it currently has a bug with dragging connections between green endpoints, caused by the fact that those endpoints use a Bezier connector and the bezier library is, due to package.json, mismatched with the core library.

- in github now on the demo repository that issue is fixed but it hasn't, at the time of writing, been deployed.

- in both this repository and the demo repository, the defaults for the jsplumb instance look like this:

```javascript
var instance = window.j = jsPlumbBrowserUI.newInstance({
            dragOptions: { cursor: 'pointer', zIndex: 2000 },
            paintStyle: { stroke: '#666'},
            endpointHoverStyle: { fill: "orange" },
            hoverPaintStyle: { stroke: "orange" },
            endpointStyle: { width: 20, height: 16, stroke: '#666' },
            endpoint: "Rectangle",
            anchors: ["TopCenter", "TopCenter"],
            container: canvas,
            dropOptions:{activeClass:"dragActive", hoverClass:"dropHover"}
        });
```

but the individual endpoint configurations override much of this.

- when you make a connection via the console using the code you shared above, it uses the defaults shown here, and this is the output:

<img width="1060" alt="image" src="https://user-images.githubusercontent.com/262720/171820993-057cdf9e-bff2-4cbc-a68c-7a9cf6b26f34.png">

the cause of the connection not appearing to paint is that `paintStyle` does not have a `strokeWidth` value set. If you change the `paintStyle` to this:

```javascript
paintStyle: { stroke: '#666', strokeWidth:1 },
```
then you get this:

<img width="526" alt="image" src="https://user-images.githubusercontent.com/262720/171821378-a770ab6a-ba51-480d-8470-b10f2b576854.png">

- The default value for `paintStyle` is:

```javascript
paintStyle: { strokeWidth: 2, stroke: "#456" },
```

- When a user provides a `paintStyle` it overrides the default - it isn't merged with it. I think it could probably be argued that jsPlumb should ensure there is a `strokeWidth` value set when the user provides a `paintStyle`.

**Comment by xu4wang** [comment]
> 

magic.. It's working on my own app now.  The only parameter missing is the default strokeWidth setting.

Thanks @sporritt for the wonderful library jsplumb  and for your quick support to my issues.

**Comment by sporritt** [comment]
ok good. sorry for the confusion, uncovering another bug whilst trying to figure this out derailed me.  

i've raised this

https://github.com/jsplumb/jsplumb/issues/1133

to track the need for ensuring `strokeWidth` is set.

**Comment by sporritt** [comment]
fix for that pushed to 5.7.2.  not on npm yet.

**Comment by xu4wang** [comment]
yes. layout was called before connect.

I first place the nodes one by one, then add the edges.


Simon Porritt ***@***.***>于2022年6月2日 周四03:27写道：
